src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -0,0 +1,666 @@
+{
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
+
+  // Find out how much is due for the operator
+  uint64_t operator_fee = 0;
+  {
+    // This calculates the operator fee using (operator_portion / max_operator_portion) * distribution_amount but using 128 bit integer math
+    uint64_t hi, lo, resulthi, resultlo;
+    lo = mul128(sn_info.portions_for_operator, distribution_amount, &hi);
+    div128_64(hi, lo, STAKING_PORTIONS, &resulthi, &resultlo);
+    if (resulthi > 0)
+      throw std::logic_error("overflow from calculating sn operator fee");
+    operator_fee = resultlo;
+  }
+  std::vector<cryptonote::batch_sn_payment> payments;
+  // Pay the operator fee to the operator
+  if (operator_fee > 0)
+    payments.emplace_back(sn_info.operator_address, operator_fee, m_nettype);
+
+  // Pay the balance to all the contributors (including the operator again)
+  uint64_t total_contributed_to_sn = std::accumulate(sn_info.contributors.begin(), sn_info.contributors.end(), uint64_t(0), [](auto const a, auto const b){return a + b.amount;});
+
+  for (auto & contributor : sn_info.contributors)
+  {
+    // This calculates (contributor.amount / total_contributed_to_winner_sn) * (distribution_amount - operator_fee) but using 128 bit integer math
+    uint64_t hi, lo, resulthi, resultlo;
+    lo = mul128(contributor.amount, distribution_amount - operator_fee, &hi);
+    div128_64(hi, lo, total_contributed_to_sn, &resulthi, &resultlo);
+    if (resulthi > 0)
+      throw std::logic_error("overflow from calculating sn contributor reward");
+    if (resultlo > 0)
+      payments.emplace_back(contributor.address, resultlo, m_nettype);
+  }
+
+  return payments;
+}
+
+bool BlockchainSQLite::add_block(const cryptonote::block& block, const service_nodes::service_node_list::state_t& service_nodes_state)
+{
+
+  auto block_height = get_block_height(block);
+  MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - Add block called with block " << block_height << " - debug");
+
+  // If we receive an add_block() for the genesis block assume we are clearing the database 
+  if (block_height == 0) 
+  {
+    clear_database();
+    return true;
+  }
+
+  print_database();
+
+  auto hf_version = block.major_version;
+  if (hf_version < cryptonote::network_version_19)
+  {
+    if (height > block_height)
+    {
+      MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - " << "Clearing database because db height is above block height and pre hardfork" << " - debug");
+      clear_database();
+    }
+    return update_height(block_height);
+  }
+  auto fork_height = cryptonote::get_hard_fork_heights(m_nettype, cryptonote::network_version_19);
+  MDEBUG(__FILE__ << ":" << __LINE__ << " (" << __func__ << ") TODO sean remove this - fork height " << (int64_t)fork_height.first.value_or(0) << " - debug");
+  if (block_height == fork_height.first.value_or(0))
```

Comment:
afaict this is the only one of these height checks that is necessary, and the only time that `clear_database` should be necessary as well.  Will revisit.